apps/lib/s_cb.c

```diff
@@ -353,39 +353,39 @@ int ssl_print_groups(BIO *out, SSL *s, int noshared)
     groups = app_malloc(ngroups * sizeof(int), "groups to print");
     SSL_get1_groups(s, groups);
 
-    BIO_puts(out, "Supported Elliptic Groups: ");
+    BIO_puts(out, "Supported groups: ");
     for (i = 0; i < ngroups; i++) {
         if (i)
             BIO_puts(out, ":");
         nid = groups[i];
-        /* If unrecognised print out hex version */
-        if (nid & TLSEXT_nid_unknown) {
-            BIO_printf(out, "0x%04X", nid & 0xFFFF);
-        } else {
-            /* TODO(TLS1.3): Get group name here */
-            /* Use NIST name for curve if it exists */
-            gname = EC_curve_nid2nist(nid);
-            if (gname == NULL)
-                gname = OBJ_nid2sn(nid);
-            BIO_printf(out, "%s", gname);
-        }
+        gname = SSL_groupid_to_name(s, nid);
+        if (gname == NULL)
+            gname = OBJ_nid2sn(nid);
+        /* If still unrecognised print out hex version */
+        if ((gname == NULL) && (nid & TLSEXT_nid_unknown))
```

Comment:
Should be `||` instead?

Reply:
Nope: If we have a name we want to show that (instead of the group ID).

Reply:
Then it should simply be `if (gname == NULL)`.

Reply:
That would then change the output (i.e., externally visible behaviour) of the function compared to before: If the returned NID does not have the `TLSEXT_nid_unknown` bit set (and no name has been found), it would simply (unbeknownst to the user) output the NID instead of the group_id. Until now, this case became visible by receiving the output `(nil)`. I don't know whether this can happen (having no group name _and_ `TLSEXT_nid_unknown` unset), but the additional guard made sense to me from a safety perspective. But I'd surely follow your judgement in this regard, so will change if I don't get additional feedback.

Reply:
There is simply no reason to output nil in such case. It does not provide any useful information.

Reply:
It only helped me pinpoint quickly missing support for non-built-in groups; but as I'm all for changing something that wasn't sensible, `if (gname == NULL)` it will be.

Reply:
`SSL_groupid_to_name` should work and give a sensible name for all outputs from `SSL_get1_groups` and similar functions. Therefore it should be able to handle just a nid, or a group id with the `TLSEXT_nid_unknown` bit set. So I don't think this code printing out the hex is necessary. Either `gname` is non-null in which case we print it, or alternatively it is NULL in which case an error has occurred.

Reply:
ACK. Test removed. Error (NULL) will become printed.